### PR TITLE
Non-full-chunk support

### LINKED ISSUE #399: Changes to chunks don't show up properly
**To Reproduce**
1. Install WorldEdit (or do /fill) on Java edition
2. The Bedrock user will not see the changes until they relog

**Expected behavior**
Changes should be automatically updated
**Screenshots / Videos**

**Server Version**
This server is running Paper version git-Paper-185 (MC: 1.15.2) (Implementing API version 1.15.2-R0.1-SNAPSHOT)

**Geyser Version**
1.0 Snapshot, Inventory branch

**Additional Context**

### PR DESCRIPTION
**Pull request by DoctorMacc**
This PR adds non-full chunk support if chunk caching is enabled. This has not been strongly tested: as of PRing this I have only tested it in a flat world where the biome stays the same; I have not tested CPU toll. I will not let this be merged until Jeb himself approves. (Or we adequately decide that all cases have been thoroughly tested and CPU toll has been measured)

Fixes #399.

### CODE REVIEW

**Comment by DoctorMacc** [comment]
More testing is needed, but this PR is ready to be reviewed.

**Comment by LegacyGamerHD** [comment]
CPU and ram usage is fine

**Comment by SinanGentoo** [comment]
The problem persists when WordEdit modifies the world on a large scale

**Comment by DoctorMacc** [comment]
With this PR? Can you give an example?

**Comment by SinanGentoo** [comment]
[Video](http://pan.hydrostic.com:5212/api/v3/file/get/30/12744192916829685599379967506.mp4?sign=NXA2pMO2bXiE6qdCfQA0wvOO4GIhQg6GvAUunk8bvVM%3D%3A0)

**Comment by DoctorMacc** [comment]
@SinanGentoo and this is with this PR installed?

**Comment by SinanGentoo** [comment]
> @SinanGentoo and this is with this PR installed?
yes

**Comment by SinanGentoo** [comment]
> @SinanGentoo and this is with this PR installed?

Yes

**Comment by rtm516** [comment]
We had another report of this fixing the chunk update issue https://discordapp.com/channels/613163671870242838/613169789900161035/714448037409194035

**Comment by DoctorMacc** [comment]
@SinanGentoo also: what platform are you using (standalone/Bukkit/etc.)? Is chunk caching on?

File: connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaChunkDataTranslator.java [hunk 1]
```diff
@@ -55,12 +57,15 @@ public void translate(ServerChunkDataPacket packet, GeyserSession session) {
             ChunkUtils.updateChunkPosition(session, session.getPlayerEntity().getPosition().toInt());
         }
 
-        if (packet.getColumn().getBiomeData() == null) //Non-full chunk
+        if (packet.getColumn().getBiomeData() == null && !session.getConnector().getConfig().isCacheChunks()) //Non-full chunk without chunk caching
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaChunkDataTranslator.java:60]: Maybe worth storing `session.getConnector().getConfig().isCacheChunks()` as a static variable so it doesn't need to be fetched everytime?

File: connector/src/main/java/org/geysermc/connector/utils/ChunkUtils.java [hunk 2]
```diff
@@ -64,7 +64,7 @@
      * Temporarily stores positions of BlockState values that are needed for certain block entities actively
      */
     public static final Map<Position, BlockState> CACHED_BLOCK_ENTITIES = new HashMap<>();
-
+
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/utils/ChunkUtils.java:67]: This should be changed back for a cleaner commit

**Review by rtm516 — DISMISSED** [review]

**Comment by Loony-dev** [comment]
This code works except that in the GeyserBukkitWorldManager.java file I had to redo the getBiomeDataAt function a bit

```
    @Override
    @SuppressWarnings("deprecation")
    public int[] getBiomeDataAt(GeyserSession session, int x, int z) {
        if (session.getPlayerEntity() == null) {
            return null;
        }
        int[] biomeData = new int[1024];
        ChunkSnapshot chunk = Objects.requireNonNull(Bukkit.getPlayer(session.getPlayerEntity().getUsername())).getWorld().getChunkAt(x, z).getChunkSnapshot(true, true, true);
        for (int localX = 0; localX < 16; localX = localX + 4) {
            for (int localY = 0; localY < 255; localY = localY + 4) {
                for (int localZ = 0; localZ < 16; localZ = localZ + 4) {
                    // Index is based on wiki.vg's index requirements
                    final int i = ((localY >> 2) & 63) << 4 | ((localZ >> 2) & 3) << 2 | ((localX >> 2) & 3);
                    biomeData[i] = chunk.getBiome(localX, localZ).ordinal();
                }
            }
        }
        return biomeData;
    }
```

**Comment by DoctorMacc** [comment]
@Loony-dev why did you have to make these changes? What happened otherwise?

**Comment by Loony-dev** [comment]
> @Loony-dev why did you have to make these changes? What happened otherwise?

I used Spigot and it did not accept the third parameter, so I had to do so

**Comment by rtm516** [comment]
!start-test-server

**Comment by rtm516** [comment]
!start-test-server

**Comment by rtm516** [comment]
!start-test-server

**Review by RednedEpic — APPROVED** [review]

**Pull request merged**